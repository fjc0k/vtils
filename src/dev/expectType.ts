// eslint-disable-next-line @typescript-eslint/no-unused-vars
export function expectType<X, Y extends X>() {
  expect(1).toBe(1)
}
