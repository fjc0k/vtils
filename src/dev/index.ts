/**
 * 开发工具库。
 *
 * @packageDocumentation
 */

/* istanbul ignore file */

// @index(['./**/*.ts', '!./**/*.test.*', '!**/__*'], f => `export * from '${f.path}'`)
export * from './expectType.ts'
export * from './runBenchmark.ts'
// @endindex
