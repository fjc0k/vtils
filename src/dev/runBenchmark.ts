import { formatNumber, Options, Suite } from 'benchmark'
import Table from 'cli-table3'

export function runBenchmark(
  suites: Record<string, () => any>,
  options?: Options,
) {
  const table = new Table({
    head: ['name', 'ops/sec'],
    colAligns: ['left', 'right'],
  })
  const suite = new Suite()
  let expectFastestName: string | undefined
  for (let name of Object.keys(suites)) {
    const fn = suites[name]
    if (name.endsWith('!fastest')) {
      name = name.replace(/!fastest$/, '')
      expectFastestName = name
    }
    suite.add(name, fn)
  }
  suite.on('cycle', function (event: any) {
    table.push([event.target.name, formatNumber(Math.round(event.target.hz))])
  })
  suite.on('complete', function (this: any) {
    const actualFastestName = this.filter('fastest').map('name')[0]
    if (expectFastestName != null && expectFastestName !== actualFastestName) {
      throw new Error(
        `Expect fastest is ${expectFastestName}, actual fastest is ${actualFastestName}.`,
      )
    }
    console.log(
      table
        .sort(
          (a, b) =>
            (b as any)[1].replace(/,/, '') - (a as any)[1].replace(/,/, ''),
        )
        .toString(),
    )
  })
  suite.run(options)
}
