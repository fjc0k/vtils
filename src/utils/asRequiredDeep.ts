import { RequiredDeep } from '../types'

/**
 * 将给定的值 `RequiredDeep` 化。
 *
 * @param value 值
 */
export function asRequiredDeep<T>(value: T): RequiredDeep<T> {
  return value as any
}
