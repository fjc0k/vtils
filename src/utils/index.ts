/**
 * 基础工具库。基于 {@link https://lodash.com/docs/4.17.15 | Lodash}。
 *
 * @packageDocumentation
 */

/* istanbul ignore file */

export * from 'lodash-uni'

// @index(['./**/*.ts', '!./**/*.{test,perf}.*'], f => `export * from '${f.path}'`)
export * from './asRequiredDeep'
export * from './base64'
export * from './bindEvent'
export * from './Calculator'
export * from './cartesianProduct'
export * from './chooseFile'
export * from './chunkEqual'
export * from './cloneDeepFast'
export * from './constantCase'
export * from './copyTextToClipboard'
export * from './createSubmit'
export * from './createUrlQueryString'
export * from './DataPacker'
export * from './dedent'
export * from './defaultIndexTo'
export * from './devOrProd'
export * from './EventBus'
export * from './formatBytes'
export * from './formatNumber'
export * from './GeoCoord'
export * from './getCurrentScript'
export * from './getEnvironment'
export * from './getWechatPublicAccountQrcodeUrl'
export * from './htmlToDom'
export * from './ii'
export * from './inAndroid'
export * from './inBrowser'
export * from './inDeno'
export * from './indent'
export * from './inIOS'
export * from './inMiniProgram'
export * from './inNodeJS'
export * from './inTaro'
export * from './inWechatWebView'
export * from './isBlobUrl'
export * from './isChineseIDCardNumber'
export * from './isDataUrl'
export * from './isElementVisible'
export * from './isNumeric'
export * from './isPossibleChineseMobilePhoneNumber'
export * from './isPromiseLike'
export * from './isType'
export * from './isUrl'
export * from './jsonp'
export * from './keysStrict'
export * from './loadCss'
export * from './loadResource'
export * from './LocalStoragePlus'
export * from './loopUntil'
export * from './makeEnum'
export * from './md5'
export * from './MiniProgramUrl'
export * from './move'
export * from './omitStrict'
export * from './onceMeanwhile'
export * from './parseDataUrl'
export * from './parseUrlQueryString'
export * from './pascalCase'
export * from './pickStrict'
export * from './placeKitten'
export * from './pMap'
export * from './prepareData'
export * from './readFile'
export * from './removeBlankChars'
export * from './removeEmojiChars'
export * from './removeNonWordChars'
export * from './RichUrl'
export * from './rot13'
export * from './roundTo'
export * from './run'
export * from './sampleBy'
export * from './sampleIndex'
export * from './selectDom'
export * from './signal'
export * from './swap'
export * from './toFullWidthString'
export * from './toHalfWidthString'
export * from './traverse'
export * from './TreeData'
export * from './wait'
export * from './Wechat'
// @endindex
