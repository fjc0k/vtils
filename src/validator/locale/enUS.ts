/* istanbul ignore file */
import { yup } from '../yup.js'

// ref: https://github.com/jquense/yup/blob/master/src/locale.js

export const enUS: yup.Locale = {
  mixed: {
    default: '${path} is invalid',
    required: '${path} is a required field',
    oneOf: '${path} must be one of the following values: ${values}',
    notOneOf: '${path} must not be one of the following values: ${values}',
    defined: '${path} must be defined',
    notType: ({ path, type, value, originalValue }) => {
      const isCast = originalValue != null && originalValue !== value
      const msg = [
        `${path} must be a \`${type}\` type, `,
        `but the final value was: \`${yup.printValue(value, true)}\``,
        !isCast
          ? '.'
          : ` (cast from the value \`${yup.printValue(
              originalValue,
              true,
            )}\`).`,
        value !== null
          ? ''
          : '\n If "null" is intended as an empty value be sure to mark the schema as `.nullable()`',
      ]
        .filter(Boolean)
        .join('')
      return msg
    },
  },
  string: {
    length: '${path} must be exactly ${length} characters',
    min: '${path} must be at least ${min} characters',
    max: '${path} must be at most ${max} characters',
    matches: '${path} must match the following: "${regex}"',
    email: '${path} must be a valid email',
    url: '${path} must be a valid URL',
    trim: '${path} must be a trimmed string',
    lowercase: '${path} must be a lowercase string',
    uppercase: '${path} must be a upper case string',
    chineseIDCardNumber: '${path} must be a Chinese identity card number',
    chineseMobilePhoneNumber: '${path} must be a Chinese mobile phone number',
  },
  number: {
    min: '${path} must be greater than or equal to ${min}',
    max: '${path} must be less than or equal to ${max}',
    lessThan: '${path} must be less than ${less}',
    moreThan: '${path} must be greater than ${more}',
    positive: '${path} must be a positive number',
    negative: '${path} must be a negative number',
    integer: '${path} must be an integer',
    id: '${path} must be a positive integer',
    positiveInteger: '${path} must be a positive integer',
    negativeInteger: '${path} must be a negative integer',
    nonPositive: '${path} must be a non-positive number',
    nonNegative: '${path} must be a non-negative number',
    nonPositiveInteger: '${path} must be a non-positive integer',
    nonNegativeInteger: '${path} must be a non-negative integer',
  },
  date: {
    min: '${path} field must be later than ${min}',
    max: '${path} field must be at earlier than ${max}',
  },
  boolean: {},
  object: {
    noUnknown: '${path} field has unspecified keys: ${unknown}',
  },
  array: {
    min: '${path} field must have at least ${min} items',
    max: '${path} field must have less than or equal to ${max} items',
  },
}
