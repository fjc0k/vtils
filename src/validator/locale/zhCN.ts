/* istanbul ignore file */
import { yup } from '../yup'

export const zhCN: yup.Locale = {
  mixed: {
    default: ({ label }) => `${label || '此项'}错误`,
    required: ({ label }) => `${label || '此项'}必填`,
    oneOf: ({ label, values }) =>
      `${label || '此项'}必须是下列值之一: ${values}`,
    notOneOf: ({ label, values }) =>
      `${label || '此项'}必须不是下列值之一: ${values}`,
    defined: ({ label }) => `${label || '此项'}必须已定义`,
    notType: ({ label, type }) =>
      `${label || '此项'}必须是一个${
        type === 'number'
          ? '数字'
          : type === 'string'
          ? '字符串'
          : type === 'boolean'
          ? '布尔值'
          : type === 'object'
          ? '对象'
          : type === 'array'
          ? '数组'
          : type === 'date'
          ? '日期'
          : ` ${type} 类型的值`
      }`,
  },
  string: {
    length: ({ label, length }) =>
      `${label || '此项'}必须仅包含${length}位字符`,
    min: ({ label, min }) => `${label || '此项'}至少应包含${min}位字符`,
    max: ({ label, max }) => `${label || '此项'}最多可包含${max}位字符`,
    matches: ({ label, regex }) =>
      `${label || '此项'}必须匹配该正则表达式: ${regex}`,
    email: ({ label }) => `${label || '此项'}必须是一个合法的邮箱`,
    url: ({ label }) =>
      `${label || '此项'}必须是一个以 http:// 或 https:// 开头的网络链接`,
    trim: ({ label }) => `${label || '此项'}必须是一个没有旁白的字符串`,
    lowercase: ({ label }) => `${label || '此项'}必须是一个小写的字符串`,
    uppercase: ({ label }) => `${label || '此项'}必须是一个大写的字符串`,
    chineseMobilePhoneNumber: ({ label }) =>
      `${label || '此项'}必须是一个合法的手机号码`,
    chineseIDCardNumber: ({ label }) =>
      `${label || '此项'}必须是一个合法的身份证号码`,
  },
  number: {
    min: ({ label, min }) => `${label || '此项'}必须大于或等于${min}`,
    max: ({ label, max }) => `${label || '此项'}必须小于或等于${max}`,
    lessThan: ({ label, less }) => `${label || '此项'}必须小于${less}`,
    moreThan: ({ label, more }) => `${label || '此项'}必须大于${more}`,
    positive: ({ label }) => `${label || '此项'}必须是一个正数`,
    negative: ({ label }) => `${label || '此项'}必须是一个负数`,
    integer: ({ label }) => `${label || '此项'}必须是一个整数`,
    id: ({ label }) => `${label || '此项'}必须是一个正整数`,
  },
  date: {
    min: ({ label, min }) => `${label || '此项'}必须晚于${min}`,
    max: ({ label, max }) => `${label || '此项'}必须早于${max}`,
  },
  boolean: {},
  object: {
    noUnknown: ({ label, unknown }) =>
      `${label || '此项'}上有未定义的键: ${unknown}`,
  },
  array: {
    min: ({ label, min }) => `${label || '此项'}至少应包含${min}个元素`,
    max: ({ label, max }) => `${label || '此项'}最多可包含${max}个元素`,
  },
}
