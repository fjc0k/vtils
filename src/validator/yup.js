import { zhCN } from './locale/zhCN'
import * as yup from './yupSource'

// 设置中文为默认语言
yup.setLocale(zhCN)

export { yup }
