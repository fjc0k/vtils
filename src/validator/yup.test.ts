import { expectType } from '../dev'
import { run } from '../utils'
import { enUS } from './locale/enUS'
import { zhCN } from './locale/zhCN'
import { yup } from './yup'

describe('yup', () => {
  test('默认语言为中文', async () => {
    await expect(
      yup
        .string()
        // @ts-expect-error
        .validate(1, {
          strict: true,
        }),
    ).rejects.toThrowError(/此项必须是一个字符串/)
  })

  test('可切换语言', () => {
    yup.setLocale(enUS)
    yup.setLocale(zhCN)
  })

  test('chineseMobilePhoneNumber 正常', async () => {
    await expect(
      yup.string().chineseMobilePhoneNumber().validate('10086'),
    ).rejects.toThrowError(/此项必须是一个合法的手机号码/)
    await expect(
      yup.string().chineseMobilePhoneNumber().validate('13907199856'),
    ).resolves.toBe('13907199856')
  })

  test('chineseIDCardNumber 正常', async () => {
    await expect(
      yup.string().chineseIDCardNumber().validate('123456'),
    ).rejects.toThrowError(/此项必须是一个合法的身份证号码/)
    await expect(
      yup.string().chineseIDCardNumber().validate('11010119881101331X'),
    ).resolves.toBe('11010119881101331X')
  })

  test('validateInOrder 正常', async () => {
    const rule = yup.object({
      name: yup.string().label('姓名').required(),
      pass: yup.number().label('密码').required().min(100),
    })
    const [nameErr] = await run(() =>
      rule.validateInOrder({
        name: '',
        pass: 99,
      }),
    )
    expect(nameErr).toMatchSnapshot('姓名必填')
    const [passErr] = await run(() =>
      rule.validateInOrder({
        pass: 99,
        name: '',
      }),
    )
    expect(passErr).toMatchSnapshot('密码有误')
    expect(
      await rule.validateInOrder({
        name: '8',
        pass: 101,
      }),
    ).toEqual({
      name: '8',
      pass: 101,
    })
  })

  test('validateInOrderSync 正常', async () => {
    const rule = yup.object({
      name: yup.string().label('姓名').required(),
      pass: yup.number().label('密码').required().min(100),
    })
    const [nameErr] = await run(() =>
      rule.validateInOrderSync({
        name: '',
        pass: 99,
      }),
    )
    expect(nameErr).toMatchSnapshot('姓名必填')
    const [passErr] = await run(() =>
      rule.validateInOrderSync({
        pass: 99,
        name: '',
      }),
    )
    expect(passErr).toMatchSnapshot('密码有误')
    expect(
      rule.validateInOrderSync({
        name: '8',
        pass: 101,
      }),
    ).toEqual({
      name: '8',
      pass: 101,
    })
  })

  test('validatePlus 正常', async () => {
    const rule = yup.object({
      name: yup.string().label('姓名').required(),
      pass: yup.number().label('密码').required().min(100),
    })
    expect(
      await rule.validatePlus({
        name: '',
        pass: 99,
      }),
    ).toMatchSnapshot()
    expect(
      await rule.validatePlus({
        pass: 99,
        name: '',
      }),
    ).toMatchSnapshot()
    expect(
      await rule.validatePlus({
        name: '8',
        pass: 101,
      }),
    ).toMatchSnapshot()
    expect(
      await rule.validatePlus({
        name: '8',
        pass: 101,
        // @ts-ignore
        xxx: 'sss',
      }),
    ).toMatchSnapshot()
  })

  test('validatePlusSync 正常', async () => {
    const rule = yup.object({
      name: yup.string().label('姓名').required(),
      pass: yup.number().label('密码').required().min(100),
    })
    expect(
      rule.validatePlusSync({
        name: '',
        pass: 99,
      }),
    ).toMatchSnapshot()
    expect(
      rule.validatePlusSync({
        pass: 99,
        name: '',
      }),
    ).toMatchSnapshot()
    expect(
      rule.validatePlusSync({
        name: '8',
        pass: 101,
      }),
    ).toMatchSnapshot()
    expect(
      rule.validatePlusSync({
        name: '8',
        pass: 101,
        // @ts-ignore
        xxx: 'sss',
      }),
    ).toMatchSnapshot()
  })

  test('布尔、enum类型正常', () => {
    type X = yup.GetObjectSchema<{
      disabled: boolean
      gender: 'male' | 'female'
    }>
    expectType<
      X,
      {
        disabled: yup.BooleanSchema<boolean>
        gender: yup.StringSchema<'male' | 'female'>
      }
    >()
  })

  test('test 第一个参数可传函数或正则', () => {
    const rule = yup.object({
      name: yup.string().label('姓名').required().test(/f/, '姓名应该包含f'),
      pass: yup
        .string()
        .label('密码')
        .required()
        .test(value => value.length > 10, '密码长度应大于10'),
    })
    expect(
      rule.validatePlusSync({
        name: 'de',
        pass: 'dss',
      }),
    ).toMatchSnapshot()
    expect(
      rule.validatePlusSync({
        name: 'defl',
        pass: 'dss',
      }),
    ).toMatchSnapshot()
    expect(
      rule.validatePlusSync({
        name: 'defl',
        pass: 'dssdllsal;;',
      }),
    ).toMatchSnapshot()
  })

  test('number 支持 id()', () => {
    const rule = yup.number().id()
    expect(rule.validatePlusSync(1)).toMatchSnapshot()
    expect(rule.validatePlusSync(-1)).toMatchSnapshot()
    expect(rule.validatePlusSync(0)).toMatchSnapshot()
    expect(rule.validatePlusSync(0.5)).toMatchSnapshot()
    expect(rule.validatePlusSync(1.5)).toMatchSnapshot()
    expect(rule.validatePlusSync(100)).toMatchSnapshot()
  })

  test('number 支持 positiveInteger()', () => {
    const rule = yup.number().positiveInteger()
    expect(rule.validatePlusSync(1)).toMatchSnapshot()
    expect(rule.validatePlusSync(-1)).toMatchSnapshot()
    expect(rule.validatePlusSync(0)).toMatchSnapshot()
    expect(rule.validatePlusSync(0.5)).toMatchSnapshot()
    expect(rule.validatePlusSync(1.5)).toMatchSnapshot()
    expect(rule.validatePlusSync(100)).toMatchSnapshot()
  })

  test('number 支持 negativeInteger()', () => {
    const rule = yup.number().negativeInteger()
    expect(rule.validatePlusSync(1)).toMatchSnapshot()
    expect(rule.validatePlusSync(-1)).toMatchSnapshot()
    expect(rule.validatePlusSync(-0)).toMatchSnapshot()
    expect(rule.validatePlusSync(0.5)).toMatchSnapshot()
    expect(rule.validatePlusSync(1.5)).toMatchSnapshot()
    expect(rule.validatePlusSync(100)).toMatchSnapshot()
    expect(rule.validatePlusSync(-100.3)).toMatchSnapshot()
  })

  test('number 支持 nonPositive()', () => {
    const rule = yup.number().nonPositive()
    expect(rule.validatePlusSync(1)).toMatchSnapshot()
    expect(rule.validatePlusSync(-1)).toMatchSnapshot()
    expect(rule.validatePlusSync(-0)).toMatchSnapshot()
    expect(rule.validatePlusSync(0.5)).toMatchSnapshot()
    expect(rule.validatePlusSync(1.5)).toMatchSnapshot()
    expect(rule.validatePlusSync(100)).toMatchSnapshot()
    expect(rule.validatePlusSync(-100.3)).toMatchSnapshot()
  })

  test('number 支持 nonNegative()', () => {
    const rule = yup.number().nonNegative()
    expect(rule.validatePlusSync(1)).toMatchSnapshot()
    expect(rule.validatePlusSync(-1)).toMatchSnapshot()
    expect(rule.validatePlusSync(-0)).toMatchSnapshot()
    expect(rule.validatePlusSync(0.5)).toMatchSnapshot()
    expect(rule.validatePlusSync(1.5)).toMatchSnapshot()
    expect(rule.validatePlusSync(100)).toMatchSnapshot()
    expect(rule.validatePlusSync(-100.3)).toMatchSnapshot()
  })

  test('number 支持 nonPositiveInteger()', () => {
    const rule = yup.number().nonPositiveInteger()
    expect(rule.validatePlusSync(1)).toMatchSnapshot()
    expect(rule.validatePlusSync(-1)).toMatchSnapshot()
    expect(rule.validatePlusSync(-0)).toMatchSnapshot()
    expect(rule.validatePlusSync(0.5)).toMatchSnapshot()
    expect(rule.validatePlusSync(1.5)).toMatchSnapshot()
    expect(rule.validatePlusSync(100)).toMatchSnapshot()
    expect(rule.validatePlusSync(-100.3)).toMatchSnapshot()
  })

  test('number 支持 nonNegativeInteger()', () => {
    const rule = yup.number().nonNegativeInteger()
    expect(rule.validatePlusSync(1)).toMatchSnapshot()
    expect(rule.validatePlusSync(-1)).toMatchSnapshot()
    expect(rule.validatePlusSync(-0)).toMatchSnapshot()
    expect(rule.validatePlusSync(0.5)).toMatchSnapshot()
    expect(rule.validatePlusSync(1.5)).toMatchSnapshot()
    expect(rule.validatePlusSync(100)).toMatchSnapshot()
    expect(rule.validatePlusSync(-100.3)).toMatchSnapshot()
  })

  test('ref 支持 map', () => {
    const rule = yup.number().min(yup.ref('.', value => value + 1))
    expect(rule.validatePlusSync(2)).toMatchSnapshot()

    const rule2 = yup.number().min(
      yup.ref('.', {
        map: value => value + 1,
      }),
    )
    expect(rule2.validatePlusSync(2)).toMatchSnapshot()
  })

  test('allowEmptyString 正常', () => {
    const rule1 = yup.number()
    // @ts-expect-error
    expect(rule1.validatePlusSync('')).toMatchSnapshot()

    const rule2 = yup.number().allowEmptyString()
    // @ts-expect-error
    expect(rule2.validatePlusSync('')).toMatchSnapshot()
  })

  test('GetSchema', () => {
    type User = {
      id: number
      name: string
      role: 'normal' | 'vip'
      posts: Array<{
        id: number
        content: string
        image?: string
        tags: string[]
        createdAt: Date
        isTop: boolean
      }>
    }
    const _ = yup
    const userSchema: yup.GetSchema<User> = _.object({
      id: _.number().required(),
      name: _.string().required(),
      role: _.string($ => $.required()),
      posts: _.array($ =>
        $.of(
          _.object({
            id: _.number().required(),
            content: _.string().required(),
            image: _.string(),
            tags: _.array($ => $.of(_.string()).required()),
            createdAt: _.date().required(),
            isTop: _.boolean().required(),
          }),
        ),
      ),
    })
    expect(
      userSchema.validatePlusSync({
        id: 1,
        name: 'ddd',
        role: 'vip',
        posts: [
          {
            id: 1,
            content: 'dd',
            // @ts-expect-error
            createdAt: new Date('2023-07-14T03:23:28.926Z').toISOString(),
            isTop: true,
            tags: [],
          },
        ],
      }),
    ).toMatchSnapshot()
    expect(
      userSchema.validatePlusSync({
        // @ts-expect-error
        id: '1',
        name: 'ddd',
        role: 'vip',
        posts: [],
      }),
    ).toMatchSnapshot()
    expect(
      userSchema.validatePlusSync({
        // @ts-expect-error
        id: false,
        name: 'ddd',
        role: 'vip',
        posts: [],
      }),
    ).toMatchSnapshot()
  })

  test('ref root', () => {
    type Login = {
      pass1: string
      pass2: string
      v: {
        p1: string
        p2: string
      }
    }
    const _ = yup
    const loginSchema: yup.GetSchema<Login> = _.object($ =>
      $.shape({
        pass1: _.string().required(),
        pass2: _.string().required(),
        v: _.object({
          p1: _.string()
            .required()
            .equals([_.ref('@pass1')]),
          p2: _.string().required(),
        }).required(),
      }),
    )
    expect(
      loginSchema.validatePlusSync({
        pass1: '123',
        pass2: '456',
        v: {
          p1: '1234',
          p2: '4566',
        },
      }),
    ).toMatchSnapshot()
    expect(
      loginSchema.validatePlusSync({
        pass1: '123',
        pass2: '456',
        v: {
          p1: '123',
          p2: '4566',
        },
      }),
    ).toMatchSnapshot()
  })

  test('oneOf-obj', () => {
    const Status = {
      pending: 'pending',
      success: 'success',
      failed: 'failed',
    } as const
    type Status = typeof Status[keyof typeof Status]

    type Message = {
      id: number
      status: Status
    }

    const _ = yup
    const messageSchema: yup.GetSchema<Message> = _.object($ =>
      $.shape({
        id: _.number().required(),
        status: _.string($ => $.oneOf(Status)),
      }),
    )
    expect(
      messageSchema.validatePlusSync({
        id: 1,
        status: 'pending',
      }),
    ).toMatchSnapshot()
    expect(
      messageSchema.validatePlusSync({
        id: 1,
        // @ts-expect-error
        status: 'pending2',
      }),
    ).toMatchSnapshot()
  })

  test('oneOf-enum', () => {
    enum Status {
      pending = 'pending',
      success = 'success',
      failed = 'failed',
    }

    type Message = {
      id: number
      status: Status
    }

    const _ = yup
    const messageSchema: yup.GetSchema<Message> = _.object($ =>
      $.shape({
        id: _.number().required(),
        status: _.string($ => $.enum(Status)),
      }),
    )
    expect(
      messageSchema.validatePlusSync({
        id: 1,
        status: Status.pending,
      }),
    ).toMatchSnapshot()
    expect(
      messageSchema.validatePlusSync({
        id: 1,
        // @ts-expect-error
        status: 'pending2',
      }),
    ).toMatchSnapshot()
  })
})
