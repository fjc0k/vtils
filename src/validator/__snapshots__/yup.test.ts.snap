// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`yup test 第一个参数可传函数或正则 1`] = `
Object {
  "data": Object {
    "name": "de",
    "pass": "dss",
  },
  "error": [ValidationError: 姓名应该包含f],
}
`;

exports[`yup test 第一个参数可传函数或正则 2`] = `
Object {
  "data": Object {
    "name": "defl",
    "pass": "dss",
  },
  "error": [ValidationError: 密码长度应大于10],
}
`;

exports[`yup test 第一个参数可传函数或正则 3`] = `
Object {
  "data": Object {
    "name": "defl",
    "pass": "dssdllsal;;",
  },
}
`;

exports[`yup validateInOrder 正常: 姓名必填 1`] = `[ValidationError: 姓名必填]`;

exports[`yup validateInOrder 正常: 密码有误 1`] = `[ValidationError: 密码必须大于或等于100]`;

exports[`yup validateInOrderSync 正常: 姓名必填 1`] = `[ValidationError: 姓名必填]`;

exports[`yup validateInOrderSync 正常: 密码有误 1`] = `[ValidationError: 密码必须大于或等于100]`;

exports[`yup validatePlus 正常 1`] = `
Object {
  "data": Object {
    "name": "",
    "pass": 99,
  },
  "error": [ValidationError: 姓名必填],
}
`;

exports[`yup validatePlus 正常 2`] = `
Object {
  "data": Object {
    "name": "",
    "pass": 99,
  },
  "error": [ValidationError: 密码必须大于或等于100],
}
`;

exports[`yup validatePlus 正常 3`] = `
Object {
  "data": Object {
    "name": "8",
    "pass": 101,
  },
}
`;

exports[`yup validatePlus 正常 4`] = `
Object {
  "data": Object {
    "name": "8",
    "pass": 101,
    "xxx": "sss",
  },
}
`;

exports[`yup validatePlusSync 正常 1`] = `
Object {
  "data": Object {
    "name": "",
    "pass": 99,
  },
  "error": [ValidationError: 姓名必填],
}
`;

exports[`yup validatePlusSync 正常 2`] = `
Object {
  "data": Object {
    "name": "",
    "pass": 99,
  },
  "error": [ValidationError: 密码必须大于或等于100],
}
`;

exports[`yup validatePlusSync 正常 3`] = `
Object {
  "data": Object {
    "name": "8",
    "pass": 101,
  },
}
`;

exports[`yup validatePlusSync 正常 4`] = `
Object {
  "data": Object {
    "name": "8",
    "pass": 101,
    "xxx": "sss",
  },
}
`;
